lib/rubocop/cop/rspec/scattered_setup.rb

```diff
@@ -7,6 +7,14 @@ module RSpec
       #
       # Unify `before`, `after`, and `around` hooks when possible.
       #
+      # Autocorrection is not supported when `around` is used because it can
+      # be difficult to unify them.
+      #
+      # @safety
+      #   This cop's autocorrection is unsafe because it could change the order
```

Comment:
Does it also mean that the inspection can raise offences in an unsafe way?

Maybe we can fix that by checking if there are `let!`s between hooks?

Reply:
> Maybe we can fix that by checking if there are `let!`s between hooks?

`let!` is just one example. In fact, the existence of other method calls that may define the same kind of hook in between, such as `around`, `include_context`, `include`, or something may change the order of execution. In this case, autocorrect may succeed or fail. (Success here means that the test succeeds even if autocorrected)

We can choose not to autocorrect if there are such method calls in between, but I am afraid that if we do so, the number of cases where autocorrection can be applied will be considerably reduced. I think it is rare that the order could cause a problem in fact, and in many cases, the users would want it to be autocorrected even if it has a risk. This is because the users can easily check if it is wrong by executing the autocorrected test cases.

Reply:
Would it be safe to combine `before` hooks that follow one another with nothing in between?

Something like `ScatteredLet` could take care of `before` statements intermingled with other statements, and would work together quite nicely.

Reply:
I see. So if we prepare another `SafeAutoCorrect: false` cop (e.g. `RSpec/ScatteredHook`) that arranges the order of hooks so that they are next to each other, we can make `RSpec/ScatteredSetup`'s autocorrection safe.

```ruby
# bad
before { before1 }
foo
before { before2 }

# good
before { before1 }
before { before2 }
foo
```

In that case, would it be natural to change `RSpec/ScatteredSetup` so that only adjacent hooks are offensed?

```ruby
# good
before { before1 }
foo
before { before2 }
```

Reply:
> change RSpec/ScatteredSetup so that only adjacent hooks are offensed

I believe so.
@bquorning @Darhazer WDYT?

Reply:
> another `SafeAutoCorrect: false` cop (e.g. `RSpec/ScatteredHook`) that arranges the order of hooks so that they are next to each other

What I find the most appealing is that this cop doesn't have to be `Safe: false`, `SafeAutocorrect: false` is sufficient, as it suggests the right thing. And order-dependency of `let!` and hooks is evil.